### ISSUE TITLE
Pynome Naming Convention Parsing Issues

### ISSUE THREAD

**Issue by feltus** [biggstd/pynome#2]
Hi.  I was working with Colorado State to move the iRODs genome data into an NDN network when I realized there are some substantial naming/parsing issues::: 

0. In the naming specification in 

It says that genome files will be in this directory format:
[genus]_[species]{_[infraspecific name]}/[assembly_name]/

that contains these files:
[genus]\_[species]{\_[infraspecific name]}-[assembly_name].{hisat2 extension}
[genus]\_[species]{\_[infraspecific name]}-[assembly_name].gff3
[genus]\_[species]{\_[infraspecific name]}-[assembly_name].fasta
[genus]\_[species]{\_[infraspecific name]}-[assembly_name].gtf
[genus]\_[species]{\_[infraspecific name]}-[assembly_name].Splice_Sites.txt
[genus]\_[species]{\_[infraspecific name]}-[assembly_name].meta.json

However, the files are not in an "infraspecific name" subdirectory and the file names are not in the '[genus]_[species]{_[infraspecific name]}-[assembly_name].' format:

  C- /scidasZone/sysbio/genomes/Absidia_glauca
  C- /scidasZone/sysbio/genomes/Acanthamoeba_castellanii_str_neff
  C- /scidasZone/sysbio/genomes/Acidomyces_richmondensis
  C- /scidasZone/sysbio/genomes/Acidomyces_richmondensis_bfw
  C- /scidasZone/sysbio/genomes/Acremonium_chrysogenum_atcc_11550

BIG ISSUES:  

1.  There are no dashes in the file names, only underscores.  
2. What is an 'infraspecific name'?
3.  If there is a NULL level in the schema, then can we put in double underscores or dash-underscore so that when you parse the file name you always get the same number of fields between genomes.  (e.g. 'Absidia_glauca' versus 'Acremonium_chrysogenum_atcc_11550' gives different numbers of levels if you parse at an underscore and I am not sure if 'atcc' and '11550' should be infraspecific names?   Why is there no dash before the assembly name?  If there is no assembly name, then how can there be an assembly subdirectory (which doesn't seem to exist anyways).

Shall we meet to discuss these issues?

Alex

**Comment by spficklin** [comment]
Hi Alex, I removed that doc link from your message above because it wasn't protected.  But here's what I have for directory structure:

- Genome/
  - [genus]\_[species]{\_[infraspecific name]}/
    - [assembly_name]/
      - [genus]\_[species]{\_[infraspecific name]}-[assembly_name].{hisat2 extension}
      - [genus]\_[species]{\_[infraspecific name]}-[assembly_name].gff3
      - [genus]\_[species]{\_[infraspecific name]}-[assembly_name].fasta
      - [genus]\_[species]{\_[infraspecific name]}-[assembly_name].gtf
      - [genus]\_[species]{\_[infraspecific name]}-[assembly_name].Splice_Sites.txt
      - [genus]\_[species]{\_[infraspecific name]}-[assembly_name].meta.json

I just double checked the most recent results and it looks right. Here's an example for the rice dataset:

> /scidasZone/sysbio/testgenomes/Genome/Oryza_sativa/IRGSP-1:
>   Oryza_sativa-IRGSP-1.fa.gz
>   Oryza_sativa-IRGSP-1.gff3.gz
>   Oryza_sativa-IRGSP-1.gtf
>   Oryza_sativa-IRGSP-1.meta.json
>   Oryza_sativa-IRGSP-1.Splice_sites

The directory structure and naming looks correct to me.  The infra-specific name doesn't get a separate directory. 

Note, the most recent uploads are in the **/scidasZone/sysbio/testgenomesfolder**.  We should get rid of that **/scidasZone/sysbio/genomes** folder and move **/scidasZone/sysbio/testgenomesfolder/Genome** to the **/scidasZone/sysbio/Genome** directory.   @biggstd can you oblige? 

Here are answers to your questions:

1.  I checked the older **/scidasZone/sysbio/genomes** directory and even there the naming looks correct to me.  The underscores separate the taxonomic name parts and the dash separates the taxonomic name with the assembly name.  
2.  Infraspecific name is below species level (https://en.wikipedia.org/wiki/Infraspecific_name). If one does not exist then it doesn't get included.. that's the meaning behind the curly braces in the naming structure above to signify that it is optional.  
3. It's my understanding that **Acremonium_chrysogenum_atcc_11550** is the full taxonomic name (including infraspecific name).  The actual assembly for that is named **ASM76926v1**, and all of the files are properly named:  **Acremonium_chrysogenum_atcc_11550-ASM76926v1.\***.   One thing we failed to document was what to do if there are spaces in the names. Those get converted to underscores as well.  But it shouldn't be too hard to parse out. The first word is always genus, second word is always species and the rest (up to the dash) is the infraspecific name where underscores can be replaced with spaces.

**Comment by feltus** [comment]
Thanks Dr. F.!!  I understand this much better.  Is this right:

1. If there is NOT an infraspecific name, then that is known because there are two strings separated by one underscore (genus_species-) and then a dash?
2. If there are spaces in the infraspecific name, then these are known because there are two strings separated by one underscore (genus_species_) and then one or more underscores before the dash? 

Is that right?  If so, we can move on.

Go Cougs!
Alex

**Comment by spficklin** [comment]
Yes, that sounds right.

**Comment by feltus** [comment]
What is the final verdict on the naming convention?  It seems like the main issues are dashes in the infraspecific name?  Can we just convert dashes and spaces to underscores?

**Comment by spficklin** [comment]
Just confirming that we decided to leave the naming as described above?

**Comment by feltus** [comment]
Yes, but there was going to be a regular expression to parse the names.

**Comment by spficklin** [comment]
Okay, here is a regular expression that should pull out all of the components:

` ^(.*?)_(?:(.*?)_(.*?)|(.*?))-(.+)\.fa\.gz$`

This assumes that these are compressed genome files in FASTA format with a .fa.gz suffix. For other files with different file type suffix we just need to adjust the RE a bit.

If you want to turn the file list into a tab-delimited list then here is the substitution pattern for that regular expression:

`\1\t\2\4\t\3\t\5`

Here is a screenshot of a test of this RE on the files using the online https://regex101.com/ tool.

![image](https://user-images.githubusercontent.com/1719352/38631657-353d2504-3d6f-11e8-9557-e162d3ec9ddb.png)
